dcos-integration-test: use dcos_api_session in test_mesos.py

gpaul opened the pull request:
## High Level Description

In Enterprise `security:strict` mode the `mesos-master` and `mesos-agent` processes check the `Authorization` header and forbids unauthorized requests. They also supports only `https`. 

This PR fixes a test by replacing direct use of the `requests` library with the recently added `dcos_api_session` fixture that provides a similar API. Requests made using this fixture have a valid `Authorization` header injected. It also chooses the scheme (ie., `http` / `https`) appropriate to the cluster's security mode.

The initial `Connection Refused` failure was due to use of `http://` instead of `https://`. Fixing that revealed what we already knew: the `mesos-master` and `mesos-agent` processes would forbid unauthorized requests.

## Related Issues

  - [DCOS-14161](https://jira.mesosphere.com/browse/DCOS-14161) Fix `test_mesos.test_if_marathon_app_can_be_debugged` test

## Checklist for all PR's

  - [x] Included a test which will fail if code is reverted but test is not. If there is no test please explain here:
  - [x] Read the [DC/OS contributing guidelines](https://github.com/dcos/dcos/blob/master/contributing.md)
  - [x] Followed relevant code rules [Rules for Packages and Systemd](https://github.com/dcos/dcos/tree/master/docs)

## Checklist for component/package updates:

If you are changing components or packages in DC/OS (e.g. you are bumping the sha or ref of anything underneath `packages`), then in addition to the above please also include:

  - [ ] Change log from the last version integrated (this should be a link to commits for easy verification and review): [example](https://github.com/dcos/dcos-mesos-modules/compare/f6fa27d7c40f4207ba3bb2274e2cfe79b62a395a...6660b90fbbf69a15ef46d0184e36755881d6a5ae)
  - [ ] Test Results: [link to CI job test results for component]
  - [ ] Code Coverage (if available): [link to code coverage report]
___

gpaul commented:
@mellenburg @vespian @malnick @orsenthil PTAL

gpaul commented:
@mesosphere-mergebot bump-ee

gpaul commented:
Sure thing, it was blocking my exhibitor PR. I'm pretty confident that the idea of the patch is sound, but please review for coding style, etc. Thanks!

jgehrcke commented:
> This fixes a test failure where the Authorization header wasn't being provided to the mesos-master and mesos-agent processes.

Mhm. The reported symptom (a 'Connection reset by peer') cannot be addressed with adding an Authorization header. Also, from the diff it is not obvious to me what actually changed with respect to the request authentication. Is it the dynamic URL scheme handling of `dcos_api_session` which actually fixes the problem (before, we hard-coded `http` -- now it's dynamic and can be `https`)?

orsenthil commented on packages/dcos-integration-test/extra/test_mesos.py:
```diff
@@ -53,8 +38,10 @@ def test_if_marathon_app_can_be_debugged(dcos_api_session):
-        master_state_url = 'http://{}:{}/state'.format(dcos_api_session.masters[0], 5050)
-        r = requests.get(master_state_url)
+        scheme = dcos_api_session.default_url.scheme
+        master = dcos_api_session.masters[0]
```
`master_ip` to be explicit? 

Because we use the `agent_host` when it is a hostname.

gpaul replied:
Good catch. I've renamed it to `master_ip` and use it in the `dcos_api_session.get(...)` on the next two lines instead of calling `dcos_api_session.masters[0]` again.

gpaul commented:
@jgehrcke I've updated the Jira with the exact failure details. I've updated the PR description with a summary thereof. Indeed the initial `Connection Refused` error was due to our using `http` instead of `https`. Fixing that showed that we still need to send the `Authorization` header as we are then faced with 401 responses. The `dcos_api_session` fixture does exactly both these things.

jgehrcke commented:
Thanks for the update, Gustav.

> Fixing that showed that we still need to send the Authorization header as we are then faced with 401 responses.

Which authenticator changed behavior, and (why) is that expected? CC @adam-mesos

adam-mesos commented on packages/dcos-integration-test/extra/test_mesos.py:
```diff
@@ -76,9 +63,29 @@ def test_if_marathon_app_can_be_debugged(dcos_api_session):
+                json=json,
+                data=data,
+                stream=stream)
+            agent_url = '{}://{}:{}/state'.format(scheme, agent_hostname, 5051)
```
Why do you need `agent_url` here? Shouldn't you be logging the actual `url` that you're posting to?

gpaul replied:
Same reasoning as for `master_state_url`. The `url` parameter is just the request path, eg. "/state".

adam-mesos replied:
But in the case of the current callers of this function, it's `/api/v1` not `/state`, and there's no way to guarantee that's always the case since `url` is a parameter to `_post_agent` and this function can and likely will be reused elsewhere, with different urls. Please log the actual URL being 'post'ed. That was my original intention in these log lines.

jgehrcke replied:
> Please log the actual URL being 'post'ed 

Then we probably don't need to do anything here, because I think every HTTP interaction through the API object(s) is already logged accordingly.

adam-mesos replied:
Great, then we can remove these log lines. I originally put them in there for purposes of debugging test failures, but if the api_session already logs requests and responses, then it's redundant.

adam-mesos commented on packages/dcos-integration-test/extra/test_mesos.py:
```diff
@@ -76,9 +63,29 @@ def test_if_marathon_app_can_be_debugged(dcos_api_session):
+                agent_url,
+                headers,
+                json
+            )
```
Nit: Is the proper style to wrap the closing parens like this? I see it unwrapped elsewhere.

gpaul replied:
It's not, thanks I'll fix it.

gpaul replied:
Done.

adam-mesos commented on packages/dcos-integration-test/extra/test_mesos.py:
```diff
@@ -53,8 +38,10 @@ def test_if_marathon_app_can_be_debugged(dcos_api_session):
-        r = requests.get(master_state_url)
+        scheme = dcos_api_session.default_url.scheme
+        master_ip = dcos_api_session.masters[0]
+        master_state_url = '{}://{}:{}/state'.format(scheme, master_ip, 5050)
```
I don't think you need `master_state_url` (or `scheme`) either, since the line below just uses dcos_api_session's defaults with a hardcoded `/state`.

gpaul replied:
If you look at where I use `master_state_url`, you'll notice this is purely for logging. The original code logged the full URL, I decided to keep that behaviour to make this PR as non-invasive as possible.

jgehrcke commented:
>  We used to mute this OSS test in strict mode since we couldn't get it to pass without authz

Ah, I see. So it all makes sense. :ship:

gpaul commented:
@adam-mesos PTAL

adam-mesos commented on packages/dcos-integration-test/extra/test_mesos.py:
```diff
@@ -76,9 +60,21 @@ def test_if_marathon_app_can_be_debugged(dcos_api_session):
-        agent_v1_url = 'http://{}:{}/api/v1'.format(agent_hostname, 5051)
         logging.debug('Located %s with containerID %s on agent %s', app_id, container_id, agent_hostname)
 
+        # Wrapper to post, log, and validate return status
```
Nit: s/post, log,/post to agent/

gpaul replied:
I dropped the comment - it's a two-line function that is clearer than the comment itself.
